server/teamshandler_test.go

```diff
@@ -0,0 +1,349 @@
+	})
+
+	t.Run("mismatch between signer fingerprint and long keyID in signature", func(t *testing.T) {
+		mismatchedFingerprint := exampledata.ExampleFingerprint2 // 2 is in the API
+		assert.ErrorIsNil(t, err)
+
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &mismatchedFingerprint)
+		assertStatusCode(t, http.StatusBadRequest, response.Code)
+		assertHasJSONErrorDetail(t, response.Body,
+			"signature verification failed")
+	})
+
+	t.Run("dont have public key that signed the roster", func(t *testing.T) {
+		keyNotInAPI, err := pgpkey.LoadFromArmoredEncryptedPrivateKey(
+			exampledata.ExamplePrivateKey3, "test3")
+		fingerprintNotInAPI := keyNotInAPI.Fingerprint()
+		assert.ErrorIsNil(t, err)
+
+		signature, err := makeArmoredDetachedSignature([]byte(goodRoster), keyNotInAPI)
+		assert.ErrorIsNil(t, err)
+
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: signature,
+		}
+
+		response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &fingerprintNotInAPI)
+		assertStatusCode(t, http.StatusBadRequest, response.Code)
+		assertHasJSONErrorDetail(t, response.Body,
+			"public key that signed the roster has not been uploaded")
+	})
+
+	t.Run("roster signature invalid", func(t *testing.T) {
+		modifiedRoster := goodRoster + "\n# roster that's been modified"
+
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               modifiedRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &signerFingerprint)
+		assertStatusCode(t, http.StatusBadRequest, response.Code)
+		assertHasJSONErrorDetail(t, response.Body, "signature verification failed")
+	})
+
+	t.Run("invalid roster", func(t *testing.T) {
+
+		emailAddressTwice := `
+uuid = "972dce6a-3532-11e9-b8df-f32e04ceb372"
+
+[[person]]
+email = "test4@example.com"
+fingerprint = "BB3C 44BF 188D 56E6 35F4  A092 F73D 2F05 33D7 F9D6"
+
+[[person]]
+email = "test4@example.com"
+fingerprint = "BBBB BBBB BBBB BBBB BBBB  BBBB BBBB BBBB BBBB BBBB"
+`
+
+		fingerprintTwice := `
+uuid = "9ca4a81e-3532-11e9-9145-a3dd42867f07"
+
+[[person]]
+email = "test4@example.com"
+fingerprint = "BB3C 44BF 188D 56E6 35F4  A092 F73D 2F05 33D7 F9D6"
+
+[[person]]
+email = "b@example.com"
+fingerprint = "BB3C 44BF 188D 56E6 35F4  A092 F73D 2F05 33D7 F9D6"
+`
+
+		signingKeyNotInRoster := `
+uuid = "344672d2-35e8-11e9-ade3-93c56fb48f08"
+
+[[person]]
+email = "test4@example.com"
+fingerprint = "AAAA AAAA AAAA AAAA AAAA  AAAA AAAA AAAA AAAA AAAA"
+`
+
+		unverifiedEmailInRoster := `
+uuid = "113ab4ba-35e8-11e9-8aa6-f721c118df12"
+
+[[person]]
+email = "unverified-email@example.com"
+fingerprint = "BB3C 44BF 188D 56E6 35F4  A092 F73D 2F05 33D7 F9D6"
+`
+
+		invalidRosterTests := []struct {
```

Comment:
👍 These are handy and I think they'd be good to use in `fluidkeys`'s test suite too! Should I just copy and paste them?

Reply:
Hmm it's a good point: most of these tests should just rely on `team.Validate()`. You could copy those into the `team` module. That said, I don't think there's any harm leaving them in here too.

The only one that must live only in the API is the "unverified email" test, you won't be able to copy that one :)

Reply:
👍